### ISSUE TITLE
NAMO's performance gain may be attributed to weight decay implementation difference

### ISSUE THREAD

**Issue by rielfisu27856-debug** [minxin-zhg/namo#1]
## Summary

I reproduced NAMO's results and found that **NAMO's weight decay differs from standard AdamW/Muon**. When corrected to match AdamW-style decay, NAMO's performance degrades significantly. Moreover, reducing Muon's weight decay improves early convergence (though it diverges later). This suggests NAMO's improvement over Muon may stem from weight decay coefficient mismatch rather than the algorithmic innovation.

---

## The Weight Decay Difference

**NAMO (paper implementation):**
```python
adaptive_lr = (lr * math.sqrt(bc2) / (bc1 + 1e-12)) * (gn_m / v_t)
p.data.mul_(1.0 - wd * adaptive_lr_clamped)  #  namo.py
```

**Standard AdamW/Muon style:**
```python
p.data.mul_(1.0 - wd * lr )  
```

Since the adaptive lr is smaller than lr, this results in **weaker weight decay** in NAMO than comapred to Adam and Muon under the same weight decay coefficient.

---

## Experimental Results

**Setup:** GPT-2 Small (124M) on OpenWebText, ~50k iterations, matching all hyperparameters from the paper. We compared NAMO with AdamW-style decay and Muon with a reduced weight decay coefficient (0.2X) with the cases in the paper. 

| Variant | Val Loss @ 34k iter | Notes |
|---------|---------------------|-------|
| NAMO (paper) | **3.031** | Baseline |
| NAMO (AdamW-style decay) | **3.222** | +0.19 worse |
| Muon (standard decay) | **3.042** | Slightly worse than NAMO |
| Muon (reduced decay) | **7.44** | Better early, diverges after ~15k |

<img width="2083" height="740" alt="Image" src="https://github.com/user-attachments/assets/dcefaba4-8633-40e5-b0bd-2edadef3bcef" />

---

## Key Observations

1. **NAMO with AdamW-style decay degrades significantly** (+0.19 val loss), indicating the weight decay implementation is critical to NAMO's performance.

2. **Muon with reduced decay shows better early convergence** but eventually diverges, suggesting weaker regularization helps initially but hurts stability.

---

## Questions for Authors

1. Have you compared NAMO and Muon using matched effective weight decay? Is there any ablation study that rules out the effect of different weight decay levels?

---

**Comment by minxin-zhg** [comment]
Thank you for raising this issue and sharing your findings. 

In our paper, we implemented NAMO with adaptive weight decay, using Muon's default weight decay coefficient (`wd=0.1`). However, because NAMO and Muon use different learning rates, the effective value of the weight decay coefficient can differ between the two. 

To address this, we performed an ablation study comparing three variants:
1. **NAMO** with adaptive decay
2. **NAMO** with non-adaptive (AdamW-style) decay
3. **Muon** with non-adaptive (AdamW-style) decay

## Experimental Setup:
* **Model & Data:** GPT-2 Small (124M) on OpenWebText.
* **Duration:** 50K iterations.
* **Learning Rates:** Fixed at the default value for each optimizer (`lr=0.01` for NAMO, `lr=0.001` for Muon).
* **Weight Decay Sweep:** We swept `wd` across {0, 0.01, 0.03, 0.1, 0.3, 1}. 
* **Metric:** The optimal `wd` was chosen based on the lowest validation loss after 50K iterations.

## Results:

| Optimizer | Optimal `wd` | Val Loss (@ 50K iter) |
| :--- | :--- | :--- |
| NAMO (adaptive decay) | 0.03 | 2.9862 |
| NAMO (AdamW-style decay) | 0 | 2.9856 |
| Muon (AdamW-style decay) | 0.03 | 3.0234 |

<img width="492" height="410" alt="Image" src="https://github.com/user-attachments/assets/97b32183-189b-41e1-b416-bf5aa01cf53d" />

## Conclusion:
These results demonstrate that NAMO with either adaptive weight decay or AdamW-style weight decay outperforms Muon. Additionally, when the weight decay coefficient is tuned, NAMO with AdamW-style decay performs slightly better than NAMO with adaptive decay.
